core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -261,14 +282,36 @@ final class ShuffleBlockFetcherIterator(
             results.put(new SuccessFetchResult(BlockId(blockId), infoMap(blockId)._2,
               address, infoMap(blockId)._1, buf, remainingBlocks.isEmpty))
             logDebug("remainingBlocks: " + remainingBlocks)
+            enqueueDeferredFetchRequestIfNecessary()
           }
         }
         logTrace(s"Got remote block $blockId after ${Utils.getUsedTimeNs(startTimeNs)}")
       }
 
       override def onBlockFetchFailure(blockId: String, e: Throwable): Unit = {
         logError(s"Failed to get block(s) from ${req.address.host}:${req.address.port}", e)
-        results.put(new FailureFetchResult(BlockId(blockId), infoMap(blockId)._2, address, e))
+        remainingBlocks -= blockId
+        val (size, mapIndex) = infoMap(blockId)
+        e match {
+          // Catching OOM and do something based on it is only a workaround for handling the
+          // Netty OOM issue, which is not the best way towards memory management. We can
+          // get rid of it when we find a way to manage Netty's memory precisely.
+
+          // Ensure the Netty memory is at least enough for serving only one block to avoid
+          // the endless retry. And since the Netty memory is shared among shuffle, rpc, etc,
+          // modules, we use "1.5" for the overhead concern.
+          case _: OutOfDirectMemoryError if PlatformDependent.maxDirectMemory() > ( 1.5 * size) =>
+            if (isNettyOOMOnShuffle.compareAndSet(false, true)) {
+              // The fetcher can fail remaining blocks in batch for the same error. So we only
+              // log the warning once to avoid flooding the logs.
+              logWarning(s"Netty OOM happens, will retry the failed blocks")
+            }
+            deferredBlocks += blockId
```

Comment:
good point! I have put it inside `ShuffleBlockFetcherIterator.this.synchronized` to ensure the thread-safe.

Reply:
Sorry @Ngone51, I made a mistake here. The updates to the `deferredBlocks` will be done by a single netty event-loop not different netty event-loop so moving it inside `ShuffleBlockFetcherIterator.this.synchronized` should not be necessary.  
When `shuffleClient.fetchBlocks(....)`  is called, it creates a channel and at that time a netty event loop gets associated with this channel. This is the event loop which will always be the one invoking  `blockFetchingListener.onFailure()` or `blockFetchingListener.onSuccess()` for all the blocks in the request. 

In the `onBlockFetchSuccess()` it uses `ShuffleBlockFetcherIterator.this.synchronized` but that is because I think it is accessing the `var isZombie` which can also be modified by the task thread.
cc @attilapiros

Reply:
That's interesting! But if you look at `RetryingBlockFetchListener`, it also has protection against thread-safe. 

I'll take another look too.

Reply:
Yeah seems like we do need it. With retries [this](https://github.com/apache/spark/blob/74b93261af29e76b9da31b1c9f20900a818d97e6/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java#L101) is executed again and  can create another channel if the existing is not active. For a different channel, a different eventloop will be associated with it but the blockFetchingListener instance is still the same.

Reply:
Thanks for bringing this.

> For a different channel, a different eventloop will be associated with it but the blockFetchingListener instance is still the same.

To clarify, I think we use different listener instances for each retry (see `fetchAllOutstanding`). Rather, I think it protects the `currentListener`.

Reply:
`blockFetchingListener` instance I mentioned is the final callback instance which is created here and this remains the same across retries.  The one that is changed in `RetryingBlockFetchListener.fetchAllOutstanding` is the `RetryingBlockFetchListener` instance itself. However,  that just forwards the callbacks to the original listener [here](https://github.com/apache/spark/blob/068b6c8be625f018d06c13af2c661d1327daa8d4/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockFetcher.java#L216).

The synchronized in `RetryingBlockFetchListener` probably is protecting the `currentListener` as you said.

Reply:
I see. I misunderstood your last comment.